src/adapter/src/catalog/open/builtin_item_migration.rs

```diff
@@ -0,0 +1,540 @@
+/// definition of global IDs will no longer be valid. At best, the current version would have to
+/// rehydrate all objects that depend on migrated items. At worst, it would panic.
+///
+/// The high level description of this approach is that we create new shards for each migrated
+/// builtin table with the new table schema, without changing the global ID. Dependent objects are
+/// not modified but now read from the new shards.
+///
+/// A detailed description of this approach follows. It's important that all of these steps are
+/// idempotent, so that we can safely crash at any point and non-upgrades turn into a no-op.
+///
+///    1. Each environment has a dedicated persist shard, called the migration shard, that allows
+///       environments to durably write down metadata while in read-only mode. The shard is a
+///       mapping of `(GlobalId, deploy_generation)` to `ShardId`.
+///    2. Collect the `GlobalId` of all migrated tables for the current deploy generation.
+///    3. Read in the current contents of the migration shard.
+///    4. Collect all the `ShardId`s from the migration shard that are not at the current
+///       `deploy_generation` or are not in the set of migrated tables.
+///       a. If they ARE NOT mapped to a `GlobalId` in the storage metadata then they are shards
+///          from an incomplete migration. Finalize them and remove them from the migration shard.
+///          Note: care must be taken to not remove the shard from the migration shard until we are
+///          sure that they will be finalized, otherwise the shard will leak.
+///       b. If they ARE mapped to a `GlobalId` in the storage metadata then they are shards from a
+///       complete migration. Remove them from the migration shard.
+///    5. Collect all the `GlobalId`s of tables that are migrated, but not in the migration shard
+///       for the current deploy generation. Generate new `ShardId`s and add them to the migration
+///       shard.
+///    6. At this point the migration shard should only logically contain a mapping of migrated
+///       table `GlobalId`s to new `ShardId`s for the current deploy generation. For each of these
+///       `GlobalId`s such that the `ShardId` isn't already in the storage metadata:
+///       a. Remove the current `GlobalId` to `ShardId` mapping from the storage metadata.
+///       b. Finalize the removed `ShardId`s.
+///       c. Insert the new `GlobalId` to `ShardId` mapping into the storage metadata.
+///
+/// This approach breaks the abstraction boundary between the catalog and the storage metadata, but
+/// these types of rare, but extremely useful, abstraction breaks is the exact reason they are
+/// co-located.
+///
+/// Since the new shards are created in read-only mode, they will be left empty and all dependent
+/// items will fail to hydrate.
+/// TODO(jkosh44) Back-fill these tables in read-only mode so they can properly hydrate.
+async fn migrate_builtin_items_0dt(
```

Comment:
Might be good to note that, yes we do get a `txn` and it looks like we're writing stuff down, but in read only mode those changes will all be discarded. The only thing we _do_ have to be careful about in `read_only` mode is how we handle the migration shard, because there we actually do writes/deletes and have to be careful what we do.

Reply:
I added a little blurb about this and added an assert to make sure that `read_only` and `txn` are what we expect them to be.